gossipd/broadcast.c

```diff
@@ -71,3 +82,88 @@ const u8 *next_broadcast(struct broadcast_state *bstate,
 	}
 	return NULL;
 }
+
+static const struct pubkey *
+pubkey_keyof(const struct pubkey *pk)
+{
+	return pk;
+}
+
+static size_t pubkey_hash(const struct pubkey *id)
+{
+	return siphash24(siphash_seed(), id, sizeof(*id));
+}
+
+HTABLE_DEFINE_TYPE(struct pubkey,
+		   pubkey_keyof,
+		   pubkey_hash,
+		   pubkey_eq,
+		   pubkey_set);
+
+static void *corrupt(const char *abortstr, const char *problem,
+		     const struct short_channel_id *scid,
+		     const struct pubkey *node_id)
+{
+	status_broken("Gossip corrupt %s %s: %s",
+		      problem, abortstr ? abortstr : "",
+		      scid ? type_to_string(tmpctx,
+					    struct short_channel_id,
+					    scid)
+		      : type_to_string(tmpctx, struct pubkey, node_id));
+	if (abortstr)
+		abort();
+	return NULL;
+}
+
+struct broadcast_state *broadcast_state_check(struct broadcast_state *b,
```

Comment:
We should definitely guard this with `DEVELOPER` since it's quadratic in the number of gossip messages.

Reply:
Oh yes, this is painful on mainnet :wink: